backend/dal/Services/Concrete/ProjectSnapshotService.cs

```diff
@@ -0,0 +1,196 @@
+            snapshot.SalesCost = originalProject.SalesCost;
+            snapshot.SaleWithLeaseInPlace = originalProject.SaleWithLeaseInPlace;
+
+            // by default, this snapshot will be compared against the previous snapshot.
+            snapshot.FromSnapshot = originalProject.Snapshots.OrderByDescending(s => s.SnapshotOn).FirstOrDefault();
+            snapshot.FromSnapshotId = snapshot.FromSnapshot?.Id;
+
+            snapshot.SnapshotOn = DateTime.UtcNow;
+
+            this.Context.ProjectSnapshots.Add(snapshot);
+            this.Context.CommitTransaction();
+
+            return snapshot;
+        }
+
+        /// <summary>
+        /// Update the specified project snapshot metadata in the datasource.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="ArgumentNullException">Argument 'snapshot' is required.</exception>
+        /// <exception cref="RowVersionMissingException">Project snapshot rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to edit snapshot.</exception>
+        /// <returns></returns>
+        public ProjectSnapshot Update(ProjectSnapshot snapshot)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.AdminProjects);
+
+            var originalSnapshot = this.Context.ProjectSnapshots
+                .Include(p => p.Project)
+                .SingleOrDefault(p => p.Id == snapshot.Id) ?? throw new KeyNotFoundException();
+
+            if (originalSnapshot.IsFinal) throw new InvalidOperationException($"Unable to update FINAL project reports.");
+            if (originalSnapshot.ProjectId != snapshot.ProjectId) throw new InvalidOperationException($"Changing the Project of a Project Report is not allowed.");
+            if (originalSnapshot.Id == snapshot.FromSnapshotId) throw new InvalidOperationException($"Cannot compare a Project Report with itself.");
+
+            // only update snapshot metadata fields unless the snapshot on date changes
+            this.Context.Entry(originalSnapshot).CurrentValues.SetValues(snapshot);
+
+            this.Context.SaveChanges();
+            this.Context.CommitTransaction();
+
+            return Get(originalSnapshot.Id);
+        }
+
+        /// <summary>
+        /// Return a copy of the current snapshot using the current project data.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <exception cref="RowVersionMissingException">Project snapshot rowversion is required.</exception>
+        /// <exception cref="KeyNotFoundException">Project snapshot does not exist.</exception>
+        /// <exception cref="NotAuthorizedException">User does not have permission to edit snapshot.</exception>
+        /// <returns></returns>
+        public ProjectSnapshot Refresh(int id)
```

Comment:
I wonder if **Generate** is a better name, as you are creating a new snapshot for the current moment in time.